File: JAVA/145 - Binary Tree Postorder Traversal.java
/*
 145. Binary Tree Postorder Traversal (H)

 Approach 1 : Iteration
 Time: O(n)
 Space: O(n)

 Approach 2 : Recursion
 Time: O(n)
 Space: O(n) 

*/

/* Approach 1 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        Stack<TreeNode> todo = new Stack<TreeNode>();
        TreeNode visited = null;
        while (!todo.empty() || root != null) {
            while (root != null) {
                todo.push(root);
                root = root.left;
            }
            TreeNode peekNode = todo.peek();
            if (peekNode.right != null && peekNode.right != visited)
                root = peekNode.right;
            else {
                todo.pop();
                res.add(peekNode.val);
                visited = peekNode;
            }
        }
        return res;
    }
}


/* Approach 2 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        DFS(root, res);
        return res;
    }
    
    public void DFS(TreeNode root, List<Integer> res) {
        if (root == null) return;
        DFS(root.left, res);
        DFS(root.right, res);
        res.add(root.val);
    }
}

File: C++/145-Binary_Tree_Postorder_Traversal.cpp
/*
 145. Binary Tree Postorder Traversal (H)

 Recursion
 O(n),O(n)

 DFS
 1.Traverse to the end of the left child.
 2.Peek parent, if parent->right exist and havn't been visited yet, move root 
 to parent->right and goto Step 1. If parent->right doesn't exist, push parent
 to result vector and mark it visited
 O(n),O(n)

*/

// Recursion
class Solution
{
public:
    vector<int> postorderTraversal(TreeNode *root)
    {
        vector<int> ret;
        postorder(root, ret);
        return ret;
    }

    void postorder(TreeNode *root, vector<int> &ret)
    {
        if (!root) return;
        postorder(root->left, ret);
        postorder(root->right, ret);
        ret.push_back(root->val);
    }
};

// DFS
class Solution
{
public:
    vector<int> postorderTraversal(TreeNode *root)
    {
        vector<int> ret;
        stack<TreeNode *> todo;
        TreeNode *visited = nullptr;
        while (!todo.empty() || root)
        {
            while (root)
            {
                todo.push(root);
                root = root->left;
            }
            TreeNode *peekNode = todo.top();
            if (peekNode->right && peekNode->right != visited)
            {
                root = peekNode->right;
            }
            else
            {
                ret.push_back(peekNode->val);
                todo.pop();
                visited = peekNode;
            }
        }
        return ret;
    }
};